tsunami/tsunami.py

```diff
@@ -98,26 +186,53 @@ def append(self, data):
         Args:
             data: A numpy array with shape (nsamples, nchannels).
         """
-        dset = self._handle['signal_data']
+        dset = self._handle[self.base_name]
         dset.resize(dset.shape[0] + data.shape[0], axis=0)
         dset[-data.shape[0] :] = data.reshape(-1, self.channels)
 
-    def read(self, start_time=None, end_time=None) -> Tuple[dict, np.ndarray]:
+        max_scale = sorted(self.scales, key=lambda x: x.scale_factor)[-1]
+        while dset.shape[0] > max_scale.scale_factor * 2:
+            max_scale = Scale(
+                name=f'{self.base_name}_{max_scale.scale_factor*self.relative_scale_factor}',
+                parent_handle=self._handle,
+                upstream_name=max_scale.name,
+                relative_scale_factor=self.relative_scale_factor,
+            )
+            self.scales.append(max_scale)
+
+        for s in self.scales:
+            s.update()
+
+    def read(
+        self, start_time: Union[int, float], end_time: Union[int, float], npoints: Optional[int] = None
+    ) -> Tuple[dict, np.ndarray]:
         """Read data from the signal.
 
         Args:
             start_time: The first time requested
             end_time: The last time requested
+            npoints: The number of points desired. Returns all available points if None.
 
         Returns:
             A tuple containing the signal info and data as a numpy array with shape (nsamples, nchannels).
         """
-        dset = self._handle['signal_data']
-        start_time = start_time or self.start_time
-        end_time = end_time or self.start_time + dset.shape[0] / self.samplerate
-        samplerate = self.samplerate
-        start_idx = int(max(0, (start_time - self.start_time) * samplerate))
-        end_idx = int(min(dset.shape[0], (end_time - self.start_time) * samplerate))
+        if npoints:
+            requested_scale_factor = (end_time - start_time) * self.samplerate / npoints
+            if requested_scale_factor < 1:
+                dset = self._handle[self.base_name]
+                scale_factor = 1
+            else:
+                for s in sorted(self.scales, key=lambda x: x.scale_factor, reverse=True):
+                    if s.scale_factor < requested_scale_factor:
+                        break
+                dset = s._handle
+                scale_factor = s.scale_factor
```

Comment:
I'm uncomfortable with how many assumptions this relies on: self.scales must not be empty, and at least one scale must have scale_factor < requested_scale_factor. Both of these things will always be true, but it's not at all obvious and that may change in the future.

Reply:
This also cannot be reach 100% code coverage since the case of iterating over empty list -> skipping from 225 to 228 never occurs.

Reply:
Having thought about it more, self.scales will always be initialized, so that's not actually a problem. The loop is also intentionally set up such that if no scales have `scale_factor < requested_scale_factor` it will use the lowest scaling available. The only issue is the coverage of the branch that can't happen, which should just be excluded from the report using `pragma: no branch` [doc](https://coverage.readthedocs.io/en/7.0.1/branch.html#structurally-partial-branches)